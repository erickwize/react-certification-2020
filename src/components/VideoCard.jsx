import React from 'react';
import styled from 'styled-components';

const VideoCard = ({ data }) => {
  const formattedDate = new Date(data.snippet.publishedAt);
  const ye = new Intl.DateTimeFormat('en', { year: 'numeric' }).format(formattedDate);
  const mo = new Intl.DateTimeFormat('en', { month: 'short' }).format(formattedDate);
  const da = new Intl.DateTimeFormat('en', { day: '2-digit' }).format(formattedDate);

  return (
    <Card data-testid={data.etag}>
      <header>
        <img src={data.snippet.thumbnails.medium.url} alt={data.snippet.title} />
        <div>{data.snippet.title}</div>
      </header>
      {data.snippet.description !== '' ? (
        <div>{data.snippet.description}</div>
      ) : (
        <div>{`${data.snippet.channelTitle} - ${da} ${mo}, ${ye}`}</div>
      )}
    </Card>
  );
};

export default VideoCard;

const Card = styled.div`
  overflow: hidden;
  border-radius: 5px;
  box-shadow: 0px 0px 15px -5px rgba(0, 0, 0, 0.75);

  header {
    position: relative;
    width: 100%;
    max-height: 200px;
    overflow: hidden;
    cursor: pointer;
    &:hover {
      img {
        width: 175%;
        object-position: -7.5em -2em;
      }
    }

    img {
      width: 150%;
      transition: 0.2s;
      object-fit: cover;
      object-position: -5em -1em;
    }

    div {
      position: absolute;
      bottom: 0;
      padding: 3%;
      background-image: linear-gradient(0deg, #000 0%, transparent 100%);
      width: 100%;
      color: white;
      text-shadow: #000 0px 0px 5px;
    }
  }

  & > div {
    padding: 3%;
    font-size: 0.7em;
  }
`;
