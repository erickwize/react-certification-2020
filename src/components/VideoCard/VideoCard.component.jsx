import React, { useState } from 'react';

import {
  VideoCardWrapper,
  Thumbnail,
  VideoTitle,
  VideoOverlay,
  VideoDescription,
  VideoInfo,
} from './VideoCard.styles';

function VideoCard({ videoData, cardClick }) {
  const [descVisible, setDescVisible] = useState(false);

  return (
    <VideoCardWrapper data-testid="yt-videocard" onClick={() => cardClick(videoData)}>
      <Thumbnail img={videoData.thumbnail} alt={videoData.title}>
        <VideoOverlay
          onMouseEnter={() => setDescVisible(true)}
          onMouseLeave={() => setDescVisible(false)}
        >
          <VideoDescription data-testid="yt-videocard-description" visible={descVisible}>
            {videoData.description || `Dale click para ver el video ${videoData.title}`}
            <VideoInfo visible={descVisible}>
              {videoData.channel} | {videoData.uploadDate}
            </VideoInfo>
          </VideoDescription>
          <VideoTitle visible={descVisible}>
            {videoData.title.length > 35 && !descVisible
              ? `${videoData.title.substring(0, 34)}...`
              : videoData.title}
          </VideoTitle>
        </VideoOverlay>
      </Thumbnail>
    </VideoCardWrapper>
  );
}

export default VideoCard;
