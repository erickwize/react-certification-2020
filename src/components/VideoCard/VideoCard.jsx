import React from 'react';
import { Link } from 'react-router-dom';
import VideoBox from './VideoCard.styled';

function VideoCard({ data, videoList }) {
  const { title, description } = data?.snippet;
  const source = data?.snippet.thumbnails.default.url;
  const videoId = data?.id?.videoId;
  return (
    <VideoBox>
      <Link
        to={{
          pathname: `/video/${videoId}`,
          data: { data, videoList },
        }}
      >
        <img src={source} alt={title} />
        <h3>{title}</h3>
        <p data-testid="paragraph">{description}</p>
      </Link>
    </VideoBox>
  );
}

export default VideoCard;
