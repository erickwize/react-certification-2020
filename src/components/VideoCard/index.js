import React from 'react'
import styled from 'styled-components'

const VideoCardStyled = styled.div`
	width: 18%;
	margin: 20px 10px;
	border-radius: 5px;
	overflow: hidden;
	@media only screen and (max-width: 1400px) {
		width: 23%;
	}
	@media only screen and (max-width: 1100px) {
		width: 30%;
	}
	@media only screen and (max-width: 768px) {
		width: 45%;
	}
	@media only screen and (max-width: 475px) {
		width: 100%;
	}
`
const VideoThumbnail = styled.img`
	width: 100%;
	height: 120px;
	object-fit: cover;
	display: block;
`
const VideoInfo = styled.div``

const VideoTitle = styled.span`
	font-size: 14px;
	font-weight: 800;
	color: #444;
	padding: 5px;
	display: block;
`
const VideoDescription = styled.p`
	color: #888;
	font-size: 12px;
	font-style: italic;
	margin: 0;
	padding: 0 5px;
`

export default function VideoCard({ item }) {
	return (
		<VideoCardStyled>
			<VideoThumbnail src={item.snippet.thumbnails.default.url} alt="thumbnail" />
			<VideoInfo>
				<VideoTitle>{item.snippet.title}</VideoTitle>
				<VideoDescription>{item.snippet.description}</VideoDescription>
			</VideoInfo>
		</VideoCardStyled>
	)
}
