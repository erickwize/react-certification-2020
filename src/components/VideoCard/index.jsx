import React from 'react';

const VideoCard = (props) => (
  <div id="{props.key}" className="max-w-sm rounded overflow-hidden shadow-lg">
    <img src={props.urlImage} alt="" className="w-full" />
    <div className="px-2 py-4">
      <div className="font-bold text-black text-left font-sans mb-2">{props.title}</div>
      <div className="text-sm text-left font-sans">{props.description}</div>
      <div className="flex justify-between bg-gray-400">
        <div className="flex-shrink md:flex-shrink-0 text-xs self-start p-1">
          {props.publishedDate}
        </div>
        <div className="md:flex-shrink-0 bg-red-500 font-serif text-xs italic text-white p-1 self-end">
          {props.channel}
        </div>
      </div>
    </div>
  </div>
);

export default VideoCard;
