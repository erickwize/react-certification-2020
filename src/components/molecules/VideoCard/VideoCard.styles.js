import styled from 'styled-components';

export const CardContainer = styled.div`
  display: block;
  flex-direction: column;
  height: fit-content;
  width: 320px;
  height: 320px;
  background-color: rgba(185, 134, 217, 0.2);
  border-radius: 6px;
  margin-bottom: 30px;
  overflow: hidden;
  box-shadow: rgba(50, 44, 78, 0.15) 0px 5px 15px 0px;
  &:hover {
    position: 'absolute';
    opacity: 0.6;
    box-shadow: rgba(254, 223, 3, 0.12) 0px 2px 4px 0px,
      rgba(254, 223, 3) 0px 2px 16px 0px;
  }
  @media screen and (max-width: 812px) {
    width: 95%;
    max-width: 320px;
    height: 320px;
  }
  cursor: not-allowed;
`;

// box-shadow: rgba(149, 157, 165, 0.2) 0px 8px 24px;

export const TextContainer = styled.div`
  margin: 0.5rem;
`;
