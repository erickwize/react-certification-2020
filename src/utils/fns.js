function random(limit) {
  return Math.floor(Math.random() * limit);
}

function getReadableDate(timeStamp) {
  const date = new Date(timeStamp);
  const options = {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  };

  return date.toLocaleDateString(undefined, options);
}

function getUsefullData(videoData) {
  const { videoId } = videoData.id;
  const {
    publishedAt,
    title,
    channelTitle,
    description,
    thumbnails,
    channelId,
  } = videoData.snippet;
  return {
    uploadDate: getReadableDate(publishedAt),
    thumbnail: thumbnails.medium.url,
    title,
    description,
    channel: channelTitle,
    channelId,
    videoId,
    link: `https://www.youtube.com/embed/${videoId}?autoplay=1&color=white`,
  };
}

export { random, getReadableDate, getUsefullData };
