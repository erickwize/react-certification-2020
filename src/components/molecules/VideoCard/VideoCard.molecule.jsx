import React from 'react';
import { CardContainer, TextContainer } from './VideoCard.styles';
import { Thumbnail, VideoInformation } from '../../atoms';

function VideoCard({ videos }) {
  return (
    <CardContainer>
      <Thumbnail bg={<img src={videos.snippet.thumbnails.medium.url} />} />
      <TextContainer>
        <VideoInformation
          title={videos.snippet.title}
          description={videos.snippet.description}
        />
      </TextContainer>
    </CardContainer>
  );
}

export default VideoCard;
