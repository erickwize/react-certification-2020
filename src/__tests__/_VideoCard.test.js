import React from 'react';
import { screen, cleanup, render, fireEvent } from '@testing-library/react';
import VideoCard from '../components/VideoCard';
import VideoMock from '../utils/mock/youtube-videos-mock.json';
import { getReadableDate, random } from '../utils/fns';

afterEach(() => {
  cleanup();
});

const MockVideoInfo = VideoMock.items[1].snippet;
MockVideoInfo.uploadDate = getReadableDate(VideoMock.items[1].snippet.publishedAt);
MockVideoInfo.thumbnail = VideoMock.items[1].snippet.thumbnails.medium.url;
MockVideoInfo.channel = VideoMock.items[1].snippet.channelTitle;

it('Test Utils: getReadableDate', () => {
  expect(MockVideoInfo.uploadDate).toBe('September 30, 2019');
});

it('Test Utils: random', () => {
  expect(typeof random(10)).toBe('number');
});

it('Should render VideoCard Info', () => {
  const videoCard = render(<VideoCard videoData={MockVideoInfo} />);
  expect(videoCard).toMatchSnapshot();
  const videoCardComponent = screen.getByTestId('yt-videocard');
  expect(videoCardComponent).toBeInTheDocument();
  expect(videoCardComponent).toHaveTextContent(
    'Follow Hector Padilla, Wizeline Director of Engineering, for a lively tour of our office. In 2018, Wizeline opened its stunning new office in Guadalajara, Jalisco, ...Wizeline | September 30, 2019Video Tour | Welcome to Wizeline G...'
  );
});

it('Handle videoCard click event', () => {
  const cardClick = jest.fn();
  render(<VideoCard videoData={MockVideoInfo} cardClick={cardClick} />);
  const videoCardComponent = screen.getByTestId('yt-videocard');
  expect(videoCardComponent).toBeInTheDocument();
  fireEvent.click(videoCardComponent);
  expect(cardClick).toHaveBeenCalledWith(MockVideoInfo);
});

it('Handle videoCard mouse event', () => {
  const cardClick = jest.fn();
  render(<VideoCard videoData={MockVideoInfo} cardClick={cardClick} />);
  const videoCardComponent = screen.getByTestId('yt-videocard');
  fireEvent.mouseEnter(videoCardComponent);
  const videoCardComponentDes = screen.getByTestId('yt-videocard-description');
  expect(videoCardComponentDes).toBeInTheDocument();
});
