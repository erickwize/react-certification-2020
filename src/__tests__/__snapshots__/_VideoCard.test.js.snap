// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render VideoCard Info 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdfBwQ kqIhYC"
        data-testid="yt-videocard"
      >
        <div
          alt="Video Tour | Welcome to Wizeline Guadalajara"
          class="sc-eCssSg dodCww"
        >
          <div
            class="sc-gsTCUz jZgBUQ"
          >
            <div
              class="sc-jSgupP hlsFfD"
              data-testid="yt-videocard-description"
            >
              Follow Hector Padilla, Wizeline Director of Engineering, for a lively tour of our office. In 2018, Wizeline opened its stunning new office in Guadalajara, Jalisco, ...
              <p
                class="sc-hKgILt femupW"
              >
                Wizeline
                 | 
                September 30, 2019
              </p>
            </div>
            <p
              class="sc-dlfnbm jElMWh"
            >
              Video Tour | Welcome to Wizeline G...
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdfBwQ kqIhYC"
      data-testid="yt-videocard"
    >
      <div
        alt="Video Tour | Welcome to Wizeline Guadalajara"
        class="sc-eCssSg dodCww"
      >
        <div
          class="sc-gsTCUz jZgBUQ"
        >
          <div
            class="sc-jSgupP hlsFfD"
            data-testid="yt-videocard-description"
          >
            Follow Hector Padilla, Wizeline Director of Engineering, for a lively tour of our office. In 2018, Wizeline opened its stunning new office in Guadalajara, Jalisco, ...
            <p
              class="sc-hKgILt femupW"
            >
              Wizeline
               | 
              September 30, 2019
            </p>
          </div>
          <p
            class="sc-dlfnbm jElMWh"
          >
            Video Tour | Welcome to Wizeline G...
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
